ruby/phase-1.md

```diff
@@ -0,0 +1,598 @@
+size (passed to the initializer as an optional `:threads` argument). Then each activity task will be
+dispatched to an available thread (via an in-memory queue). To make sure a polling worker always has
+processing capacity a polling request will only be dispatched if there is an available thread in the
+pool (in other words the pool will have no buffer).
+
+A very similar model will be used for workflow tasks, with an added complexity of wrapping each
+workflow in its own Fiber (inside a thread). This serves two purposes:
+
+1. A workflow code can be paused and resumed during the history replay
+2. Sticky workflows can be left paused while waiting for the next workflow task
+
+Given that Fibers are very lightweight this will allow us to keep many stick workflows efficiently
+in memory and manage accordingly to the worker configuration.
+
+Here's a diagram to demonstrate the whole setup:
+
+<img src="./ruby_threads.png" width="400" align="center" alt="Temporal" />
+
+### Alternatives considered
+
+We've decided against using the `Client` to initialise the `Worker` (something that Go and Python
+SDKs are doing). It seemed like a wrong abstraction and would require making many attributes public
+on the client to make sure the worker can access them (expanding the public API of a client). Using
+a configuration object solves this and opens up further optimisation opportunities (that will be
+proposed in the 2nd phase).
+
+We can potentially use the reactor pattern via Async gem. This has the benefit of giving extra
+performance, however it would drastically reduce the performance unless activities are written in a
+reactor-friendly way. This can be further added as an optional approach to apps already using Async,
+but doesn't feel like a great default option.
+
+Additionally the SDK users will want to leverage multiple CPUs by spawning multiple worker
+processes. We have agreed that this part should be left out of the Ruby SDK and handled by
+developers themselves. Potentially this functionality can be added as a separate Ruby gem.
+
+
+## Activities
+
+Each activity is going to be represented by its own class. This decision comes with a few benefits:
+
+- There's no need to pollute global scope with activity methods (see "Global Scope" section)
+- There's no need to inject global context for communicating with Temporal. It will be available
+  through a private variable on the class instance (`#activity`)
+- It plays nicely with Ruby's OOP model
+- It allows for future extensibility including 3rd party extensions
+- Activity can be easily tested by injecting a test context into the class instead of the real one
+
+By default we'll use the name of the class as activity name (when communicating with the server).
+This is of course overridable in case the SDK users would like to use their own naming schema. And
+we'll use the method `#execute` as the primary entrypoint for executing user code.
```

Comment:
I've found this naming somewhat confusing in temporal-ruby vs the other SDKs. Workflow code calls `ActivityClass.execute` as a class method to start the activity, but the activity has a separate `.execute` instance method with the same name used as the entry point for executing the activity as well.

Does it make sense to separate these like the official SDKs where caller and callee have clearly distinct methods?